### PR TITLE
LRCI 2427

### PR DESCRIPTION
**Pull request by CalumR23**

### CODE REVIEW

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	public void getTotal() {
+		double totalTax = 0;
+		double totalAmount = 0;
+
+		for (Product item : _products) {
+			double itemTaxAmount;
+
+			if (!item.exempt && item.imported) {
+				itemTaxAmount = item.amount * 0.15;
```
Comment (CalumR23) [inline modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:39]: `itemTaxAmount = Math.cecil((item.amount * 0.10) * 20.0 ) /20.0;`

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			}
+			else {
+				itemTaxAmount = item.amount * 0.10;
+
+				totalTax += itemTaxAmount;
+
+				item.amount += itemTaxAmount;
+
+				totalAmount += item.amount;
+			}
```
Comment (kenjiheigel) [inline modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:67]: You already did this, but just leaving a comment for completion sake. Rework and simplify this conditional logic.

File: modules/test/jenkins-results-parser/src/main/java/tax/Product.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	public void setTaxedAmount(double newAmount) {
+		_taxedAmount = newAmount;
+	}
+
+	private double _amount;
+	private boolean _exempt;
+	private boolean _imported;
+	private String _name;
+	private int _quantity;
+	private double _taxedAmount;
```
Comment (kenjiheigel) [inline modules/test/jenkins-results-parser/src/main/java/tax/Product.java:79]: Also, already changed, but using private variables and encapsulation is the way to go for OOP.

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @author Calum Ragan
+ */
+public class Basket {
+
+	public void addItem(Product item) {
+		_products.add(item);
+	}
+
+	public void getTotal() {
```
Comment (kenjiheigel) [inline modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:31]: Rename this method. Usually any method that says `get` assumes that you're returning some value, but the signature for this method is `void`. This method calculates the total but it's not necessarily returning that as a variable.

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+					_DF.format(item.getTaxedAmount());
+
+			System.out.println(output);
+		}
+
+		System.out.println("Sales Tax : " + _DF.format(totalTax));
+		System.out.println("Total : " + _DF.format(totalAmount));
+	}
+
+	private static final DecimalFormat _DF = new DecimalFormat("0.00");
```
Comment (kenjiheigel) [inline modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:69]: rename to `_decimalFormat`. More of a Liferay thing, but typically, the name of the variable/instance should reflect the object.

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java (hunk 4 above)
Comment (CalumR23) [reply modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:31]: changed name to checkout

File: modules/test/jenkins-results-parser/src/main/java/tax/Basket.java (hunk 5 above)
Comment (CalumR23) [reply modules/test/jenkins-results-parser/src/main/java/tax/Basket.java:69]: changed to new var name

**Comment by pyoo47** [comment]
@CalumR23 Sorry for taking so long to get around to looking at this.

- I don't think the naming of the class, `Revenue` makes sense. Also the `startShopping()` method name seems strange to me because it appears to start and finish the shopping action.
- Product.java line 57: `for (String word : productArray) {` - The naming of this doesn't make sense it should be something like `for (String productName : productNames) {`. `word` is not really something I would expect to get from `productArray`

As an extended challenge, I'd like you to implement this as a unit test with the input being read in from resource files.

**Comment by pyoo47** [comment]
ci:test:sf

**Comment by pyoo47** [comment]
@CalumR23 

Source Format failures:

```
> Task :test:jenkins-results-parser:formatSource
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java (JavaParser)
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java (SourceCheck:JavaTermOrderCheck)
Missing copyright, see https://github.com/liferay/liferay-portal/blob/master/modules/util/source-formatter/src/main/resources/documentation/checks/copyright_check.markdown: /Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java (SourceCheck:CopyrightCheck)
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/tax/Product.java (JavaParser)
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/tax/Register.java (JavaParser)
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/tax/Basket.java (JavaParser)
/Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/test/java/com/liferay/jenkins/results/parser/SalesTaxTest.java (JavaParser)
Missing author: /Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java 9 (Checkstyle:MissingAuthorCheck)
java.lang.Exception: Found 2 formatting issues:
1: Missing copyright, see https://github.com/liferay/liferay-portal/blob/master/modules/util/source-formatter/src/main/resources/documentation/checks/copyright_check.markdown: /Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java (SourceCheck:CopyrightCheck)
2: Missing author: /Users/pyoo/dev/liferay-portal/pr/modules/test/jenkins-results-parser/src/main/java/com/liferay/jenkins/results/parser/BasePortalGitRepositoryJob.java 9 (Checkstyle:MissingAuthorCheck)

	at com.liferay.source.formatter.SourceFormatter.format(SourceFormatter.java:458)
	at com.liferay.source.formatter.SourceFormatter.main(SourceFormatter.java:287)

> Task :test:jenkins-results-parser:formatSource FAILED
```

**Comment by pyoo47** [comment]
ci:test:relevant

**Comment by kenjiheigel** [comment]
Final Review Meeting notes: 
* Register class should be more object oriented or the logic should be moved into the `SalesTaxTest` class. 
* For complex string parsing, use regular expressions. Was not necessary for the scope of this project, but moving forward, it's a good tool to have in your belt. You can google for tutorials and explanation. Here's also few sites that help create and test regular expressions: 
    * https://regex101.com/ 
    * https://regexr.com/

Peter noted that if you have free time and are blocked, you can revisit this if you like, and try to implement things based off this last feedback, but everyday work tasks should still be helpful in further solidifying those concepts. Good work and progression, and looks like you learned a lot since the first iteration! 

Closing as complete.

**Pull request closed without merging**